tests/test_cli_compile.py

```diff
@@ -3838,3 +3838,211 @@ def test_stdout_should_not_be_read_when_stdin_is_not_a_plain_file(
     out = runner.invoke(cli, [req_in.as_posix(), "--output-file", fifo.as_posix()])
 
     assert out.exit_code == 0, out
+
+
+@pytest.mark.parametrize(
+    (
+        "first_order_path_absolute",
+        "second_order_path_absolute",
+        "normed_output_absolute",
+    ),
+    (
+        # requirements.in + requirements2.in -> requirements2.in
+        pytest.param(False, False, False, id="both-relative"),
+        # /foo/requirements.in + /foo/requirements2.in -> /foo/requirements2.in
+        pytest.param(True, True, True, id="both-absolute"),
+        # /foo/requirements.in + requirements2.in -> /foo/requirements2.in
+        pytest.param(True, False, True, id="absolute-includes-relative"),
+        # requirements.in + /foo/requirements2.in -> requirements2.in
+        pytest.param(False, True, False, id="relative-includes-absolute"),
+    ),
+)
+def test_second_order_requirements_path_handling(
+    pip_conf,
+    runner,
+    tmp_path,
+    monkeypatch,
+    first_order_path_absolute,
+    second_order_path_absolute,
+    normed_output_absolute,
+):
+    """
+    Given nested requirements files, the internal requirements will be written
+    in the output, and it will be absolute or relative depending only on
+    whether or not the initial path was absolute or relative.
+
+    This is only accurate on pip>=24.3 , as earlier versions do not normalize
+    the `comes_from` data to absolute paths.
+    On lower pip versions, assert the previous output is produced, in which
+    `-r` inputs are preserved.
+    """
+    # in pip v24.3, new normalization will occur because `comes_from` started
+    # to be normalized to abspaths
+    pip_current_version = get_pip_version_for_python_executable(sys.executable)
+    expect_new_normalization = pip_current_version >= Version("24.3")
+
+    # setup: there are two files, the second one has an actual package name
+    req_in = tmp_path / "requirements.in"
+    req_in2 = tmp_path / "requirements2.in"
+    req_in2.write_text("small-fake-a\n")
+    # the data for the first file is written based on whether or not the
+    # second-order path is absolute
+    if second_order_path_absolute:
+        req_in.write_text(f"-r {req_in2}\n")
+    else:
+        req_in.write_text("-r ./requirements2.in\n")
+
+    # and the first order path is given on the CLI as absolute or relative
+    if first_order_path_absolute:
+        input_path = str(req_in)
+    else:
+        input_path = "requirements.in"
+
+    monkeypatch.chdir(tmp_path)
+    out = runner.invoke(
+        cli,
+        [
+            "--output-file",
+            "-",
+            "--quiet",
+            "--no-header",
+            "--no-emit-options",
+            "-r",
+            input_path,
+        ],
+    )
+    monkeypatch.undo()
```

Comment:
Would it be cleaner to use https://click.palletsprojects.com/en/stable/api/#click.testing.CliRunner.isolated_filesystem ? or `with monkeypatch.context()`?

Reply:
I like the context-manager for `monkeypatch`. `CliRunner.isolated_filesystem` puts another layer of tempdir management in the mix, which is okay but not necessary.